core/src/test/scala/integration/kafka/api/AdminClientIntegrationTest.scala

```diff
@@ -816,6 +816,87 @@ class AdminClientIntegrationTest extends IntegrationTestHarness with Logging {
     client.close()
   }
 
+  @Test
+  def testReplicaCanFetchFromLogStartOffsetAfterDeleteRecords(): Unit = {
+    val leaders = createTopic(topic, numPartitions = 1, replicationFactor = serverCount)
+    val followerIndex = if (leaders(0) != servers(0).config.brokerId) 0 else 1
+
+    def waitForFollowerLSOandLEOAfterRestart(expectedLSO: Long, expectedLEO: Long): Unit = {
+      TestUtils.waitUntilTrue(() => servers(followerIndex).replicaManager.getReplica(topicPartition) != None,
+                              "Expected follower to create replica for partition")
+
+      // wait until the follower discovers that log start offset moved beyond its HW
+      TestUtils.waitUntilTrue(() => {
+        servers(followerIndex).replicaManager.getReplica(topicPartition).get.logStartOffset == expectedLSO
+      }, s"Expected follower to discover new log start offset $expectedLSO")
+
+      TestUtils.waitUntilTrue(() => {
+        servers(followerIndex).replicaManager.getReplica(topicPartition).get.logEndOffset.messageOffset == expectedLEO
+      }, s"Expected new replica to catch up to log end offset $expectedLEO")
+    }
+
+    // we will produce to topic and delete records while one follower is down
+    killBroker(followerIndex)
+
+    client = AdminClient.create(createConfig)
+    sendRecords(producers.head, 100, topicPartition)
+
+    val result = client.deleteRecords(Map(topicPartition -> RecordsToDelete.beforeOffset(3L)).asJava)
```

Comment:
Yes, this is an integration test, but at least those tests are failing without my change. I just moved handling of the `UnexpectedAppendOffsetException` with truncation + append to Partition, so that there is no duplicate code in `ReplicaFetcherThread` and `ReplicaAlterLogDirsThread`. I can do a test that guarantees that log start offset is in the middle of the batch, but unfortunately there is no unit test for Partition class. Will try to create it, but not that simple (lots of mocking).